File: LeetCode/!0232. Implement Queue using Stacks.cpp
/*
0232. Implement Queue using Stacks

使用栈模拟队列操作。也就是说只能使用以下操作：
栈顶添加元素，栈顶访问、取出元素，获取栈大小，查看栈是否为空。

思路：用两个栈a, b模拟队列。

二刷：我之前的写法，写top和pop函数的时候，忘记把辅助栈b里面的元素再放回a，仍然通过了。是测试样例不完备吗？
*/

class MyQueue {
public:
    stack<int> a;
    stack<int> b;
    
    /** Initialize your data structure here. */
    MyQueue() {
        
    }
    
    /** Push element x to the back of queue. */
    void push(int x) {
        a.emplace(x);
    }
    
    /** Removes the element from in front of queue and returns that element. */
    int pop() {
        if (b.empty()) {
            while (!a.empty()) {
                int temp = a.top();
                a.pop();
                b.push(temp);
            }
        }
        int popValue = b.top();
        b.pop();

        // 我之前的写法，忘记把b里面的元素再放回a，仍然通过了。
        // 是测试样例不完备吗？
        while (!b.empty()) {
            int temp = b.top();
            b.pop();
            a.push(temp);
        }

        return popValue;
    }
    
    /** Get the front element. */
    int peek() {
        if (b.empty()) {
            while (!a.empty()) {
                int temp = a.top();
                a.pop();
                b.push(temp);
            }
        }
        int topValue = b.top();

        while (!b.empty()) {
            int temp = b.top();
            b.pop();
            a.push(temp);
        }

        return topValue;
    }
    
    /** Returns whether the queue is empty. */
    bool empty() {
        return a.empty();
    }
};

File: LeetCode/!0232. Implement Queue using Stacks.py
'''
0232. Implement Queue using Stacks

请你仅使用两个栈实现先入先出队列。队列应当支持一般队列的支持的所有操作（push、pop、peek、empty）。

'''

class MyQueue:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.stk = []
        self.temp = []
        self.current_queue = 0


    def push(self, x: int) -> None:
        """
        Push element x to the back of queue.
        
        1) 压入stk
        """
        self.stk.append(x)
        

    def pop(self) -> int:
        """
        Removes the element from in front of queue and returns that element.
       
        1）stk所有元素压入temp
        2）出栈一次
        3）temp中剩余元素压回stk，temp清空
        """
        while len(self.stk) > 0:
            self.temp.append(self.stk.pop())
        top = self.temp.pop()
        while len(self.temp) > 0:
            cur = self.temp.pop()
            self.stk.append(cur)
        return top
        
       
    def peek(self) -> int:
        """
        Get the front element.
        
        1）stk所有元素压入temp
        2）top一次
        3）temp元素放回stk
        4）temp清空
        """
        while len(self.stk) > 0:
            self.temp.append(self.stk.pop())
        top = self.temp[-1]
        while len(self.temp) > 0:
            self.stk.append(self.temp.pop())
        return top
        

    def empty(self) -> bool:
        """
        Returns whether the queue is empty.
        
        1）检查len(stk)
        """
        return len(self.stk) < 1
   